java-frontend/src/main/java/org/sonar/java/se/checks/NullDereferenceCheck.java

```diff
@@ -128,19 +134,41 @@ public ProgramState checkPreStatement(CheckerContext context, Tree syntaxNode) {
     return context.getState();
   }
 
-  private static ProgramState checkAssertJisNotNull(ProgramState currentState, CheckerContext context) {
-    Node parentNode = context.getNode().parent();
-    Tree previousSyntaxNode = parentNode.programPoint.syntaxTree();
-    if (!previousSyntaxNode.is(Tree.Kind.METHOD_INVOCATION)) {
-      return currentState;
-    }
-    MethodInvocationTree mit = (MethodInvocationTree) previousSyntaxNode;
-    if (!ASSERTJ_ASSERT_THAT.matches(mit)) {
-      return currentState;
+  private static boolean isIsNotNullMethod(MethodInvocationTree methodInvocation) {
+    return ASSERT_IS_NOT_NULL.anyMatch(methodInvocation)
+      // if semantic can not resolve the method, due to type inference
+      || (methodInvocation.symbol().isUnknown() && IS_NOT_NULL.equals(ExpressionUtils.methodName(methodInvocation).name()));
+  }
+
+  private static ProgramState checkAssertIsNotNull(ProgramState currentState, CheckerContext context, MethodInvocationTree methodInvocation) {
+    // search for the parent 'assertThat' method relying on syntax
+    MethodInvocationTree mit = methodInvocation;
+    while (true) {
+      ExpressionTree methodSelect = mit.methodSelect();
+      if (!methodSelect.is(Tree.Kind.MEMBER_SELECT)) {
+        return currentState;
+      }
+      ExpressionTree expression = ((MemberSelectExpressionTree) methodSelect).expression();
+      if (!expression.is(Tree.Kind.METHOD_INVOCATION)) {
+        return currentState;
+      }
+      mit = (MethodInvocationTree) expression;
+      if (ASSERT_THAT.anyMatch(mit)) {
+        break;
+      }
     }
-    SymbolicValue assertThatParameter = parentNode.programState.peekValue();
+    // Retrieved the MethodInvocationTree corresponding to the 'assertThat()' coupled with 'isNotNull()'.
+    // Now searching for corresponding Node in Exploded Graph
+    ExplodedGraph.Node node = context.getNode();
+    Tree previousSyntaxNode;
+    do {
+      node = node.parent();
+      previousSyntaxNode = node.programPoint.syntaxTree();
+    } while (!previousSyntaxNode.equals(mit));
```

Comment:
I'm scared of having infinite loop here in some corner case. Can we be more defensive here?

Reply:
Okay adding a defensive way to exit the loop.